Kudu.Services/FetchHelpers/OneDriveHelper.cs

```diff
@@ -0,0 +1,519 @@
+
+                result.Cursor = next;
+                return result;
+            }
+        }
+
+        private ChangesResult GetChanges(OneDriveModel.OneDriveItemCollection items, Dictionary<string, OneDriveModel.ItemInfo> ids, string rootUri)
+        {
+            ChangesResult result = new ChangesResult();
+
+            foreach (var item in items.value)
+            {
+                ids[item.id] = new OneDriveModel.ItemInfo
+                {
+                    name = item.name,
+                    parentId = item.parentReference.id
+                };
+
+                var path = GetPath(ids, item);
+                if (item.deleted != null)
+                {
+                    result.DeletionChanges.Add(new OneDriveModel.OneDriveChange { Path = path, IsDeleted = true });
+                }
+                else
+                {
+                    var change = new OneDriveModel.OneDriveChange
+                    {
+                        ContentUri = string.Format(CultureInfo.InvariantCulture, "{0}/items/{1}", rootUri, item.id),
+                        Path = path,
+                        IsFile = item.file != null,
+                        LastModifiedUtc = item.lastModifiedDateTime.ToUniversalTime()
+                    };
+
+                    if (change.IsFile)
+                    {
+                        result.FileChanges.Add(change);
+                    }
+                    else
+                    {
+                        result.DirectoryChanges.Add(change);
+                    }
+                }
+            }
+
+            return result;
+        }
+
+        private string GetPath(Dictionary<string, OneDriveModel.ItemInfo> ids, OneDriveModel.OneDriveItem item)
+        {
+            var path = item.name;
+            var parentId = item.parentReference.id;
+            while (true)
+            {
+                OneDriveModel.ItemInfo info;
+                if (!ids.TryGetValue(parentId, out info))
+                {
+                    return path;
+                }
+
+                path = string.Format(CultureInfo.InvariantCulture, @"{0}\{1}", info.name, path);
+                parentId = info.parentId;
+            }
+        }
+
+        private async Task<ApplyResult> ApplyChangesParallel(OneDriveModel.OneDriveChangeCollection changes, string accessToken)
+        {
+            List<Task> tasks = new List<Task>();
+            string wwwroot = _environment.WebRootPath;
+            int failureCount = 0;
+            int successCount = 0;
+
+            using (var sem = new SemaphoreSlim(initialCount: System.Environment.ProcessorCount, maxCount: System.Environment.ProcessorCount))
+            using (var filesPerSecLimiter = new RateLimiter(MaxFilesPerSeconds, TimeSpan.FromSeconds(1)))
+            {
+                foreach (OneDriveModel.OneDriveChange change in changes)
+                {
+                    tasks.Add(Task.Run(async () =>
```

Comment:
so that we can prepare all the task and kick them off in parallel. and let the Semaphore and RateLimiter to control the number of action can be done at a time.

we use same technique in [DropboxHelper as well](https://github.com/projectkudu/kudu/blob/master/Kudu.Services/FetchHelpers/DropboxHelper.cs#L262)